crates/dbsp/src/operator/neighborhood.rs

```diff
@@ -0,0 +1,696 @@
+where
+    T: BatchReader,
+{
+    fn new() -> Self {
+        Self {
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<T> Operator for NeighborhoodLocal<T>
+where
+    T: BatchReader + 'static,
+{
+    fn name(&self) -> Cow<'static, str> {
+        Cow::from("NeighborhoodLocal")
+    }
+    fn fixedpoint(&self, _scope: Scope) -> bool {
+        true
+    }
+}
+
+impl<T>
+    BinaryOperator<
+        T,
+        Option<NeighborhoodDescr<T::Key, T::Val>>,
+        OrdIndexedZSet<T::Key, T::Val, T::R>,
+    > for NeighborhoodLocal<T>
+where
+    T: BatchReader<Time = ()>,
+    T::Key: DBData,
+    T::Val: DBData,
+    T::R: DBWeight,
+{
+    fn eval<'a>(
+        &mut self,
+        input_trace: &T,
+        descr: &Option<NeighborhoodDescr<T::Key, T::Val>>,
+    ) -> OrdIndexedZSet<T::Key, T::Val, T::R> {
+        let mut cursor = input_trace.cursor();
+
+        if let Some(descr) = descr {
+            let (anchor_key, anchor_val) = &descr.anchor;
+
+            // Forward pass: locate the anchor and `decr.after`
+            // following rows.
+            let mut after = Vec::with_capacity(descr.after);
+            let mut offset = 0;
+
+            cursor.seek_keyval(anchor_key, anchor_val);
+            while cursor.keyval_valid() && offset < descr.after {
+                let w = cursor.weight();
+
+                if !cursor.weight().is_zero() {
+                    after.push(((cursor.key().clone(), cursor.val().clone()), w));
+                    offset += 1;
+                }
+                cursor.step_keyval();
+            }
+
+            // Reverse pass: find `descr.before` rows preceding the anchor.
+            cursor.fast_forward_keys();
+            cursor.fast_forward_vals();
+
+            let mut before = Vec::with_capacity(descr.before);
+            offset = 1;
+
+            cursor.seek_keyval_reverse(anchor_key, anchor_val);
+            if cursor.keyval_valid() && cursor.keyval() == (anchor_key, anchor_val) {
+                cursor.step_keyval_reverse();
+            }
+
+            while cursor.keyval_valid() && offset <= descr.before {
+                let w = cursor.weight();
+
+                if !cursor.weight().is_zero() {
+                    before.push(((cursor.key().clone(), cursor.val().clone()), w));
+                    offset += 1;
+                }
+                cursor.step_keyval_reverse();
+            }
+
+            // Assemble final result.
+            let mut builder = <<OrdIndexedZSet<_, _, _> as Batch>::Builder>::with_capacity(
+                (),
+                before.len() + after.len(),
+            );
+            for update in before.into_iter().rev() {
+                builder.push(update);
+            }
+            for update in after.into_iter() {
+                builder.push(update);
+            }
+
+            builder.done()
+        } else {
+            Batch::empty(())
+        }
+    }
+}
+
+/// Computes the neighborhood including row numbers.
```

Comment:
can't the UI do this?

Reply:
If we send the complete neighborhood to the UI, then yes, but if we stream changes to the neighborhood, it's hard or impossible without additional metadata.